ciao-deploy/deploy/workloads.go

```diff
@@ -0,0 +1,377 @@
+	Cleanup()
+}
+
+var images = []workloadDetails{
+	&baseWorkload{
+		url:       "https://download.fedoraproject.org/pub/fedora/linux/releases/24/CloudImages/x86_64/images/Fedora-Cloud-Base-24-1.2.x86_64.qcow2",
+		imageName: "Fedora Cloud Base 24-1.2",
+		extra:     true,
+		cloudInit: vmCloudInit,
+		opts: bat.WorkloadOptions{
+			Description: "Fedora test VM",
+			VMType:      "qemu",
+			FWType:      "legacy",
+			Defaults: bat.DefaultResources{
+				VCPUs: 2,
+				MemMB: 128,
+			},
+		},
+	},
+	&baseWorkload{
+		url:       "https://cloud-images.ubuntu.com/xenial/current/xenial-server-cloudimg-amd64-disk1.img",
+		imageName: "Ubuntu Server 16.04",
+		extra:     false,
+		cloudInit: vmCloudInit,
+		opts: bat.WorkloadOptions{
+			Description: "Ubuntu test VM",
+			VMType:      "qemu",
+			FWType:      "legacy",
+			Defaults: bat.DefaultResources{
+				VCPUs: 2,
+				MemMB: 256,
+			},
+		},
+	},
+	&clearWorkload{
+		wd: baseWorkload{
+			extra:     true,
+			cloudInit: vmCloudInit,
+			opts: bat.WorkloadOptions{
+				Description: "Clear Linux test VM",
+				VMType:      "qemu",
+				FWType:      "efi",
+				Defaults: bat.DefaultResources{
+					VCPUs: 2,
+					MemMB: 128,
+				},
+			},
+		},
+	},
+	&baseWorkload{
+		cloudInit: containerCloudInit,
+		opts: bat.WorkloadOptions{
+			Description: "Debian latest test container",
+			VMType:      "docker",
+			ImageName:   "debian:latest",
+			Defaults: bat.DefaultResources{
+				VCPUs: 2,
+				MemMB: 128,
+			},
+		},
+	},
+	&baseWorkload{
+		cloudInit: containerCloudInit,
+		opts: bat.WorkloadOptions{
+			Description: "Ubuntu latest test container",
+			VMType:      "docker",
+			ImageName:   "ubuntu:latest",
+			Defaults: bat.DefaultResources{
+				VCPUs: 2,
+				MemMB: 128,
+			},
+		},
+	},
+}
+
+func cleanup() {
+	var wg sync.WaitGroup
+	fmt.Printf("Cleaning up after error\n")
+	for _, wd := range images {
+		wg.Add(1)
+		go func(wd workloadDetails) {
+			wd.Cleanup()
+			wg.Done()
+		}(wd)
+	}
+	wg.Wait()
+}
+
+// CreateBatWorkloads creates all necessary workloads to run BAT
+func CreateBatWorkloads(ctx context.Context, allWorkloads bool, sshPublickey string, password string, imageCacheDir string) (errOut error) {
+	for _, wd := range images {
+		if wd.Extra() && !allWorkloads {
+			continue
+		}
+
+		if err := wd.Download(ctx, imageCacheDir); err != nil {
+			return errors.Wrap(err, "Error downloading image")
+		}
+	}
+
+	var wg sync.WaitGroup
+
+	for _, wd := range images {
+		if wd.Extra() && !allWorkloads {
+			continue
+		}
+
+		wg.Add(1)
+		go func(wd workloadDetails) {
+			if err := wd.Upload(ctx); err != nil {
+				if errOut == nil {
+					errOut = errors.Wrap(err, "Error uploading image")
+				}
+			}
+
+			if err := wd.CreateWorkload(ctx, sshPublickey, password); err != nil {
+				if errOut == nil {
+					errOut = errors.Wrap(err, "Error creating workload")
+				}
+			}
+
+			wg.Done()
+		}(wd)
+	}
+
+	wg.Wait()
+
+	defer func() {
```

Comment:
There's not much point in defering this.  If the intention is for this code to run if there's an error downloading an image it needs to be moved to the top of the function.